src/bucket/BucketList.cpp

```diff
@@ -551,22 +571,91 @@ BucketList::addBatch(Application& app, uint32_t currLedger,
 }
 
 void
-BucketList::restartMerges(Application& app, uint32_t maxProtocolVersion)
+BucketList::restartMerges(Application& app, uint32_t maxProtocolVersion,
+                          uint32_t ledgerSeq, bool useFutureBucketsFromHAS)
 {
+    std::shared_ptr<Bucket> snap;
     for (uint32_t i = 0; i < static_cast<uint32>(mLevels.size()); i++)
     {
         auto& level = mLevels[i];
         auto& next = level.getNext();
-        if (next.hasHashes() && !next.isLive())
+
+        // Determine whether to use future bucket from HAS, and restart merge
+        // with curr and snap for this level
+        auto removeShadows = spillLevelWithoutShadows(maxProtocolVersion, i);
```

Comment:
This reads a little tricky to me, I think there are 3 things here that I'm a little nervous about:

  1. `spillLevelWithoutShadows` would read better as `spillToLevelWithoutShadows`, because we're passing `i` and it's not clear if this is to or from level `i`. I think we're discussing a spill _from_ `i-1` _to_ `i`, right? It's easy for the reader to think the opposite and be off by one, read this as a spill _from_ level `i` rather than _to_ level `i`
  2. The function is phrased as a negative condition (`Without`) and then that is tested negated in the next `if` statement, which makes the control condition a little more confusing than necessary. Could you maybe change it to `bool needsShadows = spillToLevelNeedsShadows(...)` and then test it un-negated? (And of course change the helper function itself to return the opposite sense of its current behaviour).
  3. More substantially, you're passing `maxProtocolVersion` here, but treating it as `currProtocolVersion` in the helper function. Those aren't exactly synonymous concepts and I'm not 100% sure which one we actually have on hand here (and/or whether we need to have been passed _both_). See larger comment on `maxProtocolVersion` elsewhere.

Reply:
fixed